README
NAME
    CatalystX::I18N - Catalyst internationalisation (I18N) framework

SYNOPSIS
     package MyApp::Catalyst;
     use strict;
     use warnings;
     use Catalyst qw/
         +CatalystX::I18N::Role::Base
         +CatalystX::I18N::Role::GetLocale
         +CatalystX::I18N::Role::DateTime
         +CatalystX::I18N::Role::Maketext
     /; 
     # Choose only the roles you need 
     # CatalystX::I18N::Role::All is a convinient shortcut to load all available roles
 
     # Optionally also load request and response roles
     use CatalystX::RoleApplicator;
     __PACKAGE__->apply_request_class_roles(qw/CatalystX::I18N::TraitFor::Request/);
     __PACKAGE__->apply_response_class_roles(qw/CatalystX::I18N::TraitFor::Response/);
 
     # Add some I18N configuration
     __PACKAGE__->config( 
         name    => 'MyApp', 
         I18N    => {
             default_locale     => 'de_AT',
             locales            => {
                 'de'               => {
                     format_date        => 'dd.MM.yyyy',
                     format_datetime    => 'dd.MM.yyyy HH:mm',
                 },
                 'de_AT'            => {
                     inherits           => 'de',
                     timezone           => 'Europe/Vienna',
                     format_datetime    => 'dd.MM.yyyy uma HH\'e\'',
                 },
                 'de_DE'             => {
                     inherits            => 'de',
                     timezone            => 'Europe/Berlin',
                 },
             }
         },
     );

     package MyApp::Catalyst::Controller::Main;
     use strict;
     use warnings;
     use parent qw/Catalyst::Controller/;
 
     sub auto : Private {
         my ($self,$c) = @_;
         $c->get_locale(); 
         # Tries to fetch the locale from the folloing sources in the given order
         # 1. Session
         # 2. User settings
         # 3. Browser settings
         # 4. Client address
         # 5. Default locale from config
     }
 
     sub action : Local {
         my ($self,$c) = @_;
     
         $c->stash->{title} = $c->maketext('Hello world!');
         $c->stash->{location} = $c->i18n_geocode->name;
         $c->stash->{language} = $c->language;
         $c->stash->{localtime} = $c->i18n_datetime_format_date->format_datetime(DateTime->now);
     }

    If you want to load all available roles and traits you can use
    CatalystX::I18N::Role::All as a shortcut.

     package MyApp::Catalyst;
     use strict;
     use warnings;
     use Catalyst qw/
         +CatalystX::I18N::Role::All
     /;

DESCRIPTION
    CatalystX::I18N provides a comprehensive toolset for
    internationalisation (I18N) and localisation (L10N) of catalyst
    applications. This distribution consists of several modules that are
    designed to integrate seamlessly, but can be run idependently or
    replaced easily if necessary.

    *   CatalystX::I18N::Role::Base

        Basic I18N role that glues everything toghether.

    *   CatalystX::I18N::Role::Maketext

        Localize text via Locale::Maketext. Prefered over
        CatalystX::I18N::Role::DataLocalize

    *   CatalystX::I18N::Role::DataLocalize

        Localize text via Data::Localize. Alternative to
        CatalystX::I18N::Role::Maketext

    *   CatalystX::I18N::Role::PosixLocale

        Sets the POSIX locale

    *   CatalystX::I18N::Role::DateTime

        Methods for localizing date and time informations.

    *   CatalystX::I18N::Role::NumberFormat

        Methods for localizing numbers.

    *   CatalystX::I18N::TraitFor::Request

        Extends Catalyst::Request with usefull methods to help dealing with
        various I18N related information in HTTP requests.

    *   CatalystX::I18N::TraitFor::Response

        Adds a "Content-Language" header to the response.

    *   CatalystX::I18N::Role::GetLocale

        Tries to determine the most appropriate locale for the current
        request.

    *   CatalystX::I18N::Model::Maketext

        Provides access to Locale::Maketext classes via a Catalyst model.

    *   CatalystX::I18N::Model::DataLocalize

        Provides access to a Data::Localize class via a Catalyst model.

    *   CatalystX::I18N::Maketext

        Helpful wrapper around Locale::Maketext. Can also be used outside of
        Catalyst.

    *   CatalystX::I18N::DataLocalize

        Helpful wrapper around Data::Localize. Can also be used outside of
        Catalyst.

CONFIGURATION
    In order to work properly, CatalystX::I18N will need some values in your
    Catalyst configuration

     __PACKAGE__->config( 
         name    => 'MyApp', 
         I18N    => {
             default_locale     => 'de_AT',
             locales            => {
                 'de'               => {
                     inactive           => 1,
                     # Mark this locale as inactive (sort of abstract locale)
                     ...
                     # Arbitrary configuration parameters
                 },
                 'de_AT'            => {
                     inherits           => 'de',
                     # Inherit all settings form locale 'de'
                     ...
                 },
             }
         },
     );

    The configuration must be stored under the key "I18N". It should contain
    a hashref of "locales" and optionally a default locale
    ("default_locale").

    Locales can be marked as "inactive". Inactive locales will not be
    selected by the "get_locale" in CatalystX::I18N::Role::GetLocale method.

    Locales can inherit from other locales ("inherits"). All configuration
    values from inherited locales will be copied. If you use
    CatalystX::I18N::Model::Maketext together with CatalystX::I18N::Maketext
    the generated lexicons will also inherit in the given order.

    Additional configuration values are defined by the various
    CatalystX::I18N::Role::* plugins.

EXTENDING
    Extending the functionality of CatalystX::I18N distribution is easy.

    E.g. writing a new plugin that does some processing when the locale
    changes

     package CatalystX::MyI18N::Role::MyPlugin;
     use Moose::Role;
     use namespace::autoclean;
 
     after 'set_locale' => sub {
         my ($c,$locale) = @_;
         $c->do_someting($c->locale);
     };
 
     1;

SEE ALSO
    Locale::Maketext, <Locale::Maketext::Lexicon>, Data::Localize,
    Number::Format, DateTime::Locale, DateTime::Format::CLDR,
    DateTime::TimeZone, HTTP::BrowserDetect and Locale::Geocode

SUPPORT
    Please report any bugs or feature requests to
    "catalystx-i18n@rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/Public/Bug/Report.html?Queue=CatalystX::I18N>. I
    will be notified and then you'll automatically be notified of the
    progress on your report as I make changes.

AUTHOR
        Maroš Kollár
        CPAN ID: MAROS
        maros [at] k-1.com
    
        http://www.k-1.com

COPYRIGHT
    CatalystX::I18N is Copyright (c) 2012 Maroš Kollár -
    <http://www.k-1.com>

LICENCE
    This library is free software, you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl module CatalystX::I18N

1.12 Sat Apr 07 2012
    - Removed built in DateTime related type constraints in favour of global 
      type constraints (MooseX::Types::DateTime) (moltar via github)
    - Fixed failing tests (moltar via github)
    - Fix dependencies
    - Improve documentation

1.11 Sat Apr 07 2012
    - Maintenance release

1.10 Sun Jan 22 2012
    - Fix test failures on perl < 5.010
    - Renamed i18n_timezone to i18n_datetime_timezone (moltar via github)

1.09 Wed Oct 26 2011
    - Fix failing tests
    - Refactor POSIX locale handling into a separate role

1.08 Fri 14 Oct 2011
    - Add support for Data::Localize localization
    - Better checks for valid locales
    - Refactoring and code cleanup

1.07 Wed 28 Sep 2011
    - Fix broken fallback random locale (RT71098, Alexander Hosfeld)
    - Fix wrong helper filenames (RT71098, Alexander Hosfeld)

1.06 Thu 08 Sep 2011
    - Fix conflicting role method names error with latest Moose
    - Check against invalid browser languages
    - Dist cleanup

1.05 Sun 12 Dec 2010
    - Add CatalystX::I18N::TraitFor::ViewTT
    - Add CatalystX::I18N::Role::Colate
    - Rename Model::L10N an Role::L10N to *::Maketext (Incompatible change!)

1.04 Sat 16 Oct 2010
    - Added more missing dependencies
    - Add support for fail_with method in L10N model
    - Better posix setlocale support (check for return value)
    - Extend documentation

1.03 Tue 12 Oct 2010
    - Added missing dependencies
    - Fixed bugs in GetLocale role
    - Make sure that request/response traits will be applied only once
    - Documentation typos
    - Better locale checks
    - More tests

1.02 Sun 10 Oct 2010
    - Added missing dependencies
    - Better handling of default locales
    - 'All' role for loading all available roles and traits
    - Required perl version 5.010 -> 5.008

1.01 Wed 06 Oct 2010
    - Added missing dependencies
    - Better handling of default locales
    - Language-only locales are also accepted

1.00 Wed 06 Oct 2010
    - original version


Makefile.PL
# Load the Module::Install bundled in ./inc/
use strict;
use inc::Module::Install 1.05;

my $from = 'lib/CatalystX/I18N.pm';

# Define metadata
name            'CatalystX-I18N';
all_from        $from;
readme_from     $from;

perl_version    '5.008';

# Specific dependencies
requires        'Catalyst::Runtime'                     => '5.80004';
requires        'Moose'                                 => 0;
requires        'parent'                                => 0;
requires        'namespace::autoclean'                  => 0;

requires        'Path::Class'                           => 0;
requires        'MooseX::Types::Path::Class'            => 0; 
requires        'Clone'                                 => 0;

requires        'Locale::Maketext::Lexicon'             => 0;
requires        'Locale::Maketext'                      => 0;
requires        'Locale::Geocode'                       => 0;

requires        'DateTime'                              => 0;
requires        'DateTime::Locale'                      => '0.40';
requires        'DateTime::Format::CLDR'                => 0;
requires        'DateTime::TimeZone'                    => 0;
requires        'MooseX::Types::DateTime'               => 0;

requires        'IP::Country::Fast'                     => 0;
requires        'HTTP::BrowserDetect'                   => '1.09';
requires        'Number::Format'                        => 0;
requires        'Unicode::Collate'                      => 0;
requires        'Unicode::Collate::Locale'              => 0;

requires        'Data::Localize'                        => 0;
requires        'Data::Localize::Format::Gettext'       => 0;

build_requires  'Test::Most'                            => 0;
build_requires  'Test::NoWarnings'                      => 0;
build_requires  'Test::WWW::Mechanize::Catalyst'        => 0;
build_requires  'JSON'                                  => '2.17';
build_requires  'Catalyst::Plugin::Session'             => 0;
build_requires  'Catalyst::Plugin::Session::State::Cookie'=> 0;
build_requires  'Catalyst::Plugin::Session::Store::File'=> 0;
build_requires  'Catalyst::View::TT'                    => 0;
build_requires  'Template'                              => 0;

author_requires 'Test::Pod'                             => '1.14';
author_requires 'Test::Pod::Coverage'                   => '1.04';
author_requires 'Test::Perl::Critic'                    => 0;
author_requires 'Module::Install::ReadmeFromPod'        => 0;
author_requires 'Module::Install::AuthorRequires'       => 0;

repository 'git://github.com/maros/CatalystX-I18N.git';

WriteAll;